bin/run.sh

```diff
@@ -25,15 +25,42 @@ SLUG="$1"
 INPUT_DIR="${2%/}"
 OUTPUT_DIR="${3%/}"
 
-if [[ "${RUN_IN_DOCKER}" == "TRUE" ]]; then  
+if [[ "${RUN_IN_DOCKER}" == "TRUE" ]]; then
+    # Docker image contains a prebuilt package called TestEnvironment with a ready .build directory 
+    # to build solutions as fast as it possible. To minimize changes in the build graph,
+    # exercise resources are copied inside the TestEnvironment package as if they had always been there.
+
     WORKING_DIR="${PWD}"
-    cp -r "${INPUT_DIR}/." "${WORKING_DIR}"
+    CONFIG_FILE="${INPUT_DIR}/.meta/config.json"
+    SOURCE_TARGET_PATH="${WORKING_DIR}/Sources/TestEnvironment"
+    TEST_TARGET_PATH="${WORKING_DIR}/Tests/TestEnvironmentTests"
+
+    # 1. Replace source files with those of an exercise, saving TestEnvironement paths
+    rm -Rf $SOURCE_TARGET_PATH $TEST_TARGET_PATH
+
+    target_name=$(jq -r '.files.solution[0]' $CONFIG_FILE | xargs dirname | xargs basename)
+    test_target_name=$(jq -r '.files.test[0]' $CONFIG_FILE | xargs dirname | xargs basename)
```

Comment:
```suggestion
    target_name=$(jq -r '.files.solution[0]' "${CONFIG_FILE}" | xargs dirname | xargs basename)
    target_name=$(jq -r '.files.solution[0]' "${CONFIG_FILE}")
    target_name=${target_name%/*}
    target_name=${target_name##*/}
    test_target_name=$(jq -r '.files.test[0]' "$CONFIG_FILE")
    test_target_name=${test_target_name%/*}
    test_target_name=${test_target_name##*/}
```

A wrapper function can also be useful here.